### DISCUSSION TITLE
How to call a nx-plugin generator from another nx-plugin generator

### DISCUSSION THREAD

**Discussion by mehrad-rafigh** [nrwl/nx#5125] [General]
Hi there,

I am trying to migrate our workspace generators to a nx-plugin. I am currently having difficulties calling my base generator. I am wrapping it inside `wrapAngularDevkitSchematic`

Running my generator I get
Package "@cg/nx-extensions" was found but does not support schematics.

What is the way to call a generator from another generator?

**Comment by tinesoft** [comment]
Hi,

First of all, you said:
> I am wrapping it inside wrapAngularDevkitSchematic
That means your base *generator* is in fact a *schematics* ? If so, use the appropriate wording (which is schematics). 

In fact, a *generator* (nx devkit) does not require to be wrapped inside a `wrapAngularDevkitSchematic` to be called by another generator.

Now to answer your question (assuming you want to call a real nx's generator from another one), I would say 2 things are needed to have this work:
1. your to-be-called generator method must be made public/exported (inside the `index.ts ` of the generator's package)
2. its `package.json` must have a `"generators" `property that points to the `generators.json` declaring the generator (its implementation, schema, etc)

Then in your calling generator just do: `await myBaseGenerator(tree, [options])` and you should be good to go!

**Reply by mehrad-rafigh** [reply]
> That means your base generator is in fact a schematics ? If so, use the appropriate wording (which is schematics).

Since nx 11, schematics are now called generators and that's what I am building ¯\_(ツ)_/¯

> In fact, a generator (nx devkit) does not require to be wrapped inside a wrapAngularDevkitSchematic to be called by another generator.

Yeah...no. You need to wrap calls against schematics from the nx plugin with `wrapAngularDevkitSchematic`
https://github.com/nrwl/nx/issues/4869
https://nrwlcommunity.slack.com/archives/C017QSST318/p1614365826009800?thread_ts=1614345613.009400&cid=C017QSST318

**Reply by tinesoft** [reply]
Well, To be more precised, the wrapping with `wrapAngularDevkitSchematic` is necesary if your "generator" has been written and exposed with **Angular's Devkit API** (`@angular-devkit/*`). If the "generator" is already written with **Nrwl's Devkit API** (i.e with  `@nrwl/devkit`, new in Nx11),
then this schematics (angular terms) is in fact a generator (in nx11 terms) , and so, no wrapping should be necessary to called it.

Maybe if you share your code I can assist more?
